### DISCUSSION TITLE
geom_marker ggswim spin off?

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#71] [General]
hey @rsh52! Thank you again for the presentation last week!

I circled back to the geom_swim_marker question.  Having taken another look at the package, I think there could be broad interest in the data viz community - beyond the swimmer plot context.  

geom_swim_marker alliased to geom_marker  (along with scale_marker_discrete) seems like the  cool, flexible cousin of geom_point()!

I think there is some other work in this space (emoGG, others), but I'm not sure that other projects offer as in-grammar an experience.

I'm posting here so folks can weigh in about managing a potential spin off.

---

I had a few more notes... I think one way you can communicate how useful and flexible this work is, is to disconnect the glyph, color, and label dictionary from the all_events data in your README introduction.

I think people won't necessarily understand what's happening (I wasn't sure) as it is currently presented:

![image](https://github.com/user-attachments/assets/671259a9-882b-4112-9495-172d337f8a84)

If you instead write out the vectors, I think it'll be much clear what you all have accomplished 🏆🥇 (and how much more widely applicable this all might be)!! 🤯

![image](https://github.com/user-attachments/assets/353f3c8d-62e7-4ebb-8fad-bdda115e4396)

Below, trying geom_marker (geom_swim_marker alias), way out of swimmer plot context - pretty much like geom_point().

```r
library(ggplot2)
library(ggswim)
geom_marker <- geom_swim_marker

mtcars |>
  ggplot() +
  aes(x = wt, y = mpg, marker = factor(am)) +
  geom_marker(size = 8) + 
  scale_marker_discrete(
    limits = 0:1,
    colours = c("orange", "midnightblue")
  )
#> Warning in scale_marker_discrete(limits = 0:1, colours = c("orange", "midnightblue")): Continuous limits supplied to discrete scale.
#> ℹ Did you mean `limits = factor(...)` or `scale_*_continuous()`?
```

![image](https://github.com/user-attachments/assets/a95a7fb5-d54e-4f42-a90a-7fe7d93ef6d1)

Created on 2024-12-18 with [reprex v2.1.0](https://reprex.tidyverse.org/)

That all said, I think the 'getting fonts to work', will be more acute with people's desire to use your tools, in #60,  (you'll note my lack of glyphs in the 'use geom_marker() like geom_point()' example.

One final note, it seemed like it would be nicer if scale_marker should automatically detected the limits basedon aes(marker = )... I don't know if that would be possible.

**Comment by teunbrand** [comment]
I think the limits are specified here to dictate the display order, so you can depend that `glyphs` and `colour` are also for that order.

**Reply by EvaMaeRey** [reply]
Yeah, it seems like scale_color_manual is similar; I can't remember if you have to specify limits there?

**Reply by EvaMaeRey** [reply]
It looks like the scale_*_manual do not require limits - will use default values, if not provided.

**Reply by teunbrand** [reply]
Doesn't `scale_marker_discrete()` do that too?

**Reply by EvaMaeRey** [reply]
Hmmm.  So it doesn't throw an error... 

``` r
library(ggplot2)
library(ggswim)
geom_marker <- geom_swim_marker

mtcars |>
  ggplot() +
  aes(x = wt, y = mpg, marker = factor(am)) +
  geom_marker(size = 8) + 
  scale_marker_discrete(
    # limits = 0:1,
    colours = c("orange", "midnightblue")
  )
```

![](https://i.imgur.com/MQvSrSx.png)

<sup>Created on 2024-12-18 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>


Vs. declaring limits:

``` r
library(ggplot2)
library(ggswim)
geom_marker <- geom_swim_marker

mtcars |>
  ggplot() +
  aes(x = wt, y = mpg, marker = factor(am)) +
  geom_marker(size = 8) + 
  scale_marker_discrete(
    limits = 0:1,
    colours = c("orange", "midnightblue")
  )
#> Warning in scale_marker_discrete(limits = 0:1, colours = c("orange", "midnightblue")): Continuous limits supplied to discrete scale.
#> ℹ Did you mean `limits = factor(...)` or `scale_*_continuous()`?
```

![](https://i.imgur.com/ay42L7F.png)

<sup>Created on 2024-12-18 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by rsh52** [reply]
> I think the limits are specified here to dictate the display order, so you can depend that glyphs and colour are also for that order.

This is correct, I know that for my stakeholders in addition to asking for splitting of lanes and markers that people would like to see certain markers above others rather than by alphabetical order or something.

**Comment by rsh52** [comment]
Thanks @EvaMaeRey !!! I'm very interested in knowing other examples of how users might benefit from "marker" manipulation. That could really take this from being a project based in specific research asks to a larger generalizable benefit.

> If you instead write out the vectors, I think it'll be much clear what you all have accomplished 🏆🥇 (and how much more widely applicable this all might be)!! 🤯

I actually think this is a great point and probably something that could have made the presentation better as well instead of making the columns for `glyph` and `colour` a prerequisite for plotting. Total oversight on my part!

```r
p +
+     scale_marker_discrete(
+         glyphs = c("⬤", "⬤", "⚠️", "❌", "✅"),
+         colours = c("red", "orange", NA, NA, NA),
+         limits = c("First Reinfusion", "Second Reinfusion", "Other End Study Reason", "Deceased", "Completed Study Follow-Up"),
+         name = "Study Events"
+     )
```
![image](https://github.com/user-attachments/assets/f61da74b-f4a4-4a3b-86b9-4f8e11e5704a)

> That all said, I think the 'getting fonts to work', will be more acute with people's desire to use your tools, in https://github.com/teunbrand/ggplot-extension-club/discussions/60, (you'll note my lack of glyphs in the 'use geom_marker() like geom_point()' example.

Something that stuck in my head from the last few talks, like Hadley's was the purpose of a package and keeping it focused. Initially I started writing this package out of an abundance of separate requests from different users so the origins of it stem from hitting all of those. Now I fully believe the fonts issue could be separated out from {ggswim} and left up to {icons} as mentioned/authored by @mitchelloharawild (any chance this will be back on CRAN in the future?). 

I think it would be great to have that relationship built and then just point the {ggswim} documentation over to a package that's dedicated to font orchestration (plus it would selfishly offload size limits and maintenance on my end).

> One final note, it seemed like it would be nicer if scale_marker should automatically detected the limits basedon aes(marker = )... I don't know if that would be possible.

Hmm, this is a good point. I'm not sure but I can certainly investigate and it could be worth a new Issue.

-----

In general I'm curious what people think should actually fall under ggswim's ownership. I would like to get this to a point where it can be submitted to CRAN (been in the works for >1 year now!) and placed in users hands in the off chance it can get more people making swimmer plots in better and faster ways. However, I also want to be cognizant that methods in it are based in more generalizable needs of other use cases.

**Reply by mitchelloharawild** [reply]
Regarding {icons}, I agree that separating out these icon/font issues to a separate ggplot extension package is best. I can put icons on CRAN if needed, it just needs some extra work on vctrs support. I've always thought that a ggplot2 partner-package to icons would be useful, such as ggicons, which enables {icons} vectors to be added to plots (much like ggimage but more focused). Then {icons} can handle the icon management, and {ggicons} can be used for other non-ggswim applications.

If you / anyone would like to make {ggicons}, then I can reprioritise finishing {icons} and put it on CRAN.

Also noting that {icons} doesn't use fonts for the icons, but instead svg images. So there is probably some reworking of the glyphs and geoms needed (hopefully simplifications!)

**Reply by rsh52** [reply]
Thanks Mitchell! CRAN would be lovely if we go this route. 

Ah I think I misunderstood that {icons} is svg-specific. I think for this use case I would like to stick with fonts since `geom_swim_marker()` wraps `geom_text()` and the integration with the `family` arg is pretty easy to navigate. Perhaps there's an avenue somewhere with blending {ggicons}/{"ggfonts"} to ease loading fonts into users environments.
